File: frontend-food/src/schemas/aiInteraction.ts
import { z } from 'zod';

export const AiVoteInSchema = z.object({
  vote: z.enum(['up', 'down']),
});
export type AiVoteIn = z.infer<typeof AiVoteInSchema>;

export const AiVoteOutSchema = z.object({
  status: z.literal('ok'),
});
export type AiVoteOut = z.infer<typeof AiVoteOutSchema>;

export const AiContextStatsSchema = z.object({
  context: z.string(),
  label: z.string(),
  total: z.number(),
  success_count: z.number(),
  error_count: z.number(),
  thumbs_up: z.number(),
  thumbs_down: z.number(),
  vote_rate: z.number(),
  total_tokens: z.number(),
  total_cost_eur: z.number(),
});
export type AiContextStats = z.infer<typeof AiContextStatsSchema>;

export const AiTimelineEntrySchema = z.object({
  date: z.string(),
  total: z.number(),
  thumbs_up: z.number(),
  thumbs_down: z.number(),
});
export type AiTimelineEntry = z.infer<typeof AiTimelineEntrySchema>;

export const AiInteractionStatsSchema = z.object({
  total_calls: z.number(),
  calls_today: z.number(),
  voted_calls: z.number(),
  vote_rate: z.number(),
  total_tokens_all: z.number(),
  total_cost_eur: z.number(),
  by_context: z.array(AiContextStatsSchema),
  timeline: z.array(AiTimelineEntrySchema),
});
export type AiInteractionStats = z.infer<typeof AiInteractionStatsSchema>;

export const AiInteractionItemSchema = z.object({
  id: z.string().uuid(),
  context: z.string(),
  model: z.string(),
  user_name: z.string().nullable(),
  created_at: z.string(),
  total_tokens: z.number().nullable(),
  cost_eur: z.number().nullable(),
  duration_ms: z.number().nullable(),
  success: z.boolean(),
  error_code: z.string(),
  vote: z.string().nullable(),
  is_background: z.boolean(),
});
export type AiInteractionItem = z.infer<typeof AiInteractionItemSchema>;

export const AiInteractionDetailSchema = AiInteractionItemSchema.extend({
  prompt: z.union([z.record(z.unknown()), z.array(z.unknown()), z.string()]).nullable(),
  response: z.string(),
});
export type AiInteractionDetail = z.infer<typeof AiInteractionDetailSchema>;

export const UserCostSchema = z.object({
  user_id: z.number(),
  user_name: z.string(),
  total_calls: z.number(),
  total_tokens: z.number(),
  total_cost_eur: z.number(),
  cost_30d_eur: z.number(),
  vote_rate: z.number(),
});
export type UserCost = z.infer<typeof UserCostSchema>;

export const PaginatedAiInteractionsSchema = z.object({
  items: z.array(AiInteractionItemSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  total_pages: z.number(),
});
export type PaginatedAiInteractions = z.infer<typeof PaginatedAiInteractionsSchema>;

File: backend/content/schemas/ai_interaction.py
from datetime import datetime
from decimal import Decimal

from ninja import Schema


class AiVoteIn(Schema):
    vote: str  # "up" or "down"


class AiVoteOut(Schema):
    status: str = "ok"


class AiContextStatsOut(Schema):
    context: str
    label: str
    total: int
    success_count: int
    error_count: int
    thumbs_up: int
    thumbs_down: int
    vote_rate: float = 0
    total_tokens: int = 0
    total_cost_eur: float = 0


class AiTimelineEntryOut(Schema):
    date: str
    total: int
    thumbs_up: int
    thumbs_down: int
    total_cost_eur: float = 0
    total_tokens: int = 0


class AiInteractionStatsOut(Schema):
    total_calls: int
    calls_today: int
    voted_calls: int
    vote_rate: float = 0
    total_tokens_all: int = 0
    total_cost_eur: float = 0
    by_context: list[AiContextStatsOut] = []
    timeline: list[AiTimelineEntryOut] = []


# -- Admin Log Viewer Schemas --


class AiInteractionItemOut(Schema):
    id: str
    context: str
    model: str
    user_name: str | None = None
    created_at: datetime
    total_tokens: int | None = None
    cost_eur: float | None = None
    duration_ms: int | None = None
    success: bool
    error_code: str = ""
    vote: str | None = None
    is_background: bool = False


class AiInteractionDetailOut(AiInteractionItemOut):
    prompt: dict | list | str | None = None
    response: str = ""


class UserCostOut(Schema):
    user_id: int
    user_name: str
    total_calls: int
    total_tokens: int
    total_cost_eur: float
    cost_30d_eur: float
    vote_rate: float


class GeminiPricingEntryOut(Schema):
    model: str
    type: str
    input_per_1m_usd: float
    output_per_1m_usd: float | None = None
    image_output_per_1m_usd: float | None = None


class GeminiPricingOut(Schema):
    pricing: list[GeminiPricingEntryOut]
    usd_to_eur: float
